src/object/srv_obj.c

```diff
@@ -791,58 +734,235 @@ ds_obj_check_dtx_config(uint32_t flags, struct dtx_id *dti)
 	return 0;
 }
 
+/* Various check before access VOS */
+static int
+ds_pre_check(daos_unit_oid_t oid, uint32_t rpc_map_ver, uuid_t hdl_uuid,
+	     uuid_t co_uuid, uint32_t opc, uint32_t flags, struct dtx_id *dti,
+	     struct ds_cont_hdl **hdlp, struct ds_cont_child **contp)
+{
+	uint32_t	map_ver;
+	int		rc;
+
+	*hdlp = NULL;
+	*contp = NULL;
+	rc = ds_check_container(hdl_uuid, co_uuid, opc, hdlp, contp);
+	if (rc)
+		return rc;
+
+	D_ASSERT((*hdlp)->sch_pool != NULL);
+	map_ver = (*hdlp)->sch_pool->spc_map_version;
+
+	if (rpc_map_ver < map_ver) {
+		D_DEBUG(DB_IO, "stale version req %d map_version %d\n",
+			rpc_map_ver, map_ver);
+		if (obj_modify_rpc(opc))
+			D_GOTO(out_put, rc = -DER_STALE);
+		/* It is harmless if fetch with old pool map version. */
+	} else if (rpc_map_ver > map_ver) {
+		/* XXX: Client (or leader replica) has newer pool map than
+		 *	current replica. Two possibile cases:
+		 *
+		 *	1. The current replica was the old leader if with
+		 *	   the old pool map version. According to current
+		 *	   leader election algorithm, it is still the new
+		 *	   leader with the new pool map version. Since no
+		 *	   leader switch, the unmatched pool version will
+		 *	   not affect DTX related availability check.
+		 *
+		 *	2. The current replica was NOT the old leader if
+		 *	   with the old pool map version. But it becomes
+		 *	   the new leader with the new pool map version.
+		 *	   In the subsequent modificaiton, it may hit
+		 *	   some 'prepared' DTX when make availability
+		 *	   check, it will return -DER_INPROGRESS that
+		 *	   will cause client to retry. It is possible
+		 *	   that the pool map version event arrives at
+		 *	   this server during the client retry. It is
+		 *	   inefficient, but harmless.
+		 */
+	}
+
+	/* Check DTX configuration. */
+	if ((flags != 0 || dti != NULL) && obj_modify_rpc(opc) &&
```

Comment:
When the @dti can be NULL?

Reply:
hmm, right. I need update the check.

Reply:
"We need version check at here, because the pool map version may has been updated, but both the client and the leader are aware of that yet, then during the patch dispatching, as least one replica will notify the leader that the pool map has been updated."

Not sure I follow here, you mean we should check or not?  Niu and I discussed a bit, probably we should check for operation here. But in this patch, I will only check modify RPC.

Reply:
We need to check the version.